File: lib/python/src/main/opentoken/attributes/attribute_loader.py
# src/opentoken/attributes/attribute_loader.py
from typing import Set
from opentoken.attributes.attribute import Attribute
from opentoken.attributes.person.first_name_attribute import FirstNameAttribute
from opentoken.attributes.person.last_name_attribute import LastNameAttribute
from opentoken.attributes.person.birth_date_attribute import BirthDateAttribute
from opentoken.attributes.person.sex_attribute import SexAttribute
from opentoken.attributes.person.social_security_number_attribute import SocialSecurityNumberAttribute
from opentoken.attributes.person.postal_code_attribute import PostalCodeAttribute
from opentoken.attributes.general.record_id_attribute import RecordIdAttribute

class AttributeLoader:
    """Loads all available attribute implementations."""
    
    @staticmethod
    def load() -> Set[Attribute]:
        """Load all attribute implementations."""
        return {
            RecordIdAttribute(),
            FirstNameAttribute(),
            LastNameAttribute(),
            BirthDateAttribute(), 
            SexAttribute(),
            SocialSecurityNumberAttribute(),
            PostalCodeAttribute()
        }

File: lib/nodejs/opentoken/src/attributes/AttributeLoader.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { Attribute } from './Attribute';
import { FirstNameAttribute } from './person/FirstNameAttribute';
import { LastNameAttribute } from './person/LastNameAttribute';
import { BirthDateAttribute } from './person/BirthDateAttribute';
import { SexAttribute } from './person/SexAttribute';
import { SocialSecurityNumberAttribute } from './person/SocialSecurityNumberAttribute';
import { RecordIdAttribute } from './general/RecordIdAttribute';
import { StringAttribute } from './general/StringAttribute';
import { DateAttribute } from './general/DateAttribute';

/**
 * Loader for all available attributes in the system.
 * 
 * In Java, this uses ServiceLoader pattern. In Node.js/TypeScript,
 * we explicitly register attributes here.
 */
export class AttributeLoader {
  private constructor() {
    // Private constructor to prevent instantiation
  }

  /**
   * Load all available attributes.
   * 
   * @returns A Set of all available attribute instances.
   */
  static load(): Set<Attribute> {
    const attributes = new Set<Attribute>();

    // Register all attribute implementations
    attributes.add(new FirstNameAttribute());
    attributes.add(new LastNameAttribute());
    attributes.add(new BirthDateAttribute());
    attributes.add(new SexAttribute());
    attributes.add(new SocialSecurityNumberAttribute());
    attributes.add(new RecordIdAttribute());
    
    // Generic attributes (usually not used directly in token generation)
    attributes.add(new StringAttribute('String', ['String']));
    attributes.add(new DateAttribute('Date', ['Date']));

    return attributes;
  }

  /**
   * Find an attribute by name or alias.
   * 
   * @param nameOrAlias - The name or alias to search for.
   * @returns The matching attribute or undefined if not found.
   */
  static findByName(nameOrAlias: string): Attribute | undefined {
    const attributes = this.load();
    
    for (const attr of attributes) {
      if (attr.getName() === nameOrAlias || attr.getAliases().includes(nameOrAlias)) {
        return attr;
      }
    }
    
    return undefined;
  }
}
